### PR TITLE
[WIP] Wield animations

### PR DESCRIPTION
**Pull request by bendeutsch**
The shovel in minetest game points downwards when you hold it. This seems strange, but the reason becomes clear when you dig something: the digging motion is downwards, and if the shovel were the "normal" way up, it would look like the player is just hitting stuff with the flat side of the shovel.

Wouldn't it make more sense for a shovel to dig upwards? This pull request aims to do exactly that: items can define a wield animation instead of the pickaxe inspired downwards swing.

The animations are defined via splines for position (points) and rotation (quaternions), or rather spline sequences (_bonus: this pull request includes a spline sequence class!_)

The current plan is to define a standard set of wield animations (mining, digging, poking, sweeping, eating), identified by a short string, and let the item definition select this. An alternative could be for the item definition to define the splines themselves, but this is for future generations.

### CODE REVIEW

**Comment by bendeutsch** [comment]
Currently WIP; there are only two animations defined, but no way to select any yet. Still, it seemed a good idea to get this out there now to catch some feedback.

**Comment by MoNTE48** [comment]
@bendeutsch you can add video demonstration?

File: src/camera.cpp [hunk 1]
```diff
@@ -471,14 +474,64 @@ void Camera::update(LocalPlayer* player, f32 frametime, f32 busytime, f32 tool_r
 	if (m_arm_inertia)
 		addArmInertia(player->getYaw());
 
+	// New: via SplineSequences
+	std::map<std::string, SplineSequence<v3f> > positionsplines;
```
Comment (nerzhul) [inline src/camera.cpp]: i don't think order is important, unordered_map ?

**Comment by nerzhul** [comment]
i agree for a video before/after :)

File: src/camera.cpp (hunk 1 above)
Comment (bendeutsch) [reply src/camera.cpp]: True; will change.

**Comment by bendeutsch** [comment]
Videos might take a bit, for two reasons:

1. I don't have recording, editing or hosting in place yet, though I will take this as an opportunity to get something up and running.
2. This makes more sense once I have two tools with different behaviours in place, and the code's not there yet.

**Comment by LoneWolfHT** [comment]
So if I read this right. 

> The current plan is to define a standard set of wield animations (mining, digging, poking, sweeping, >eating), identified by a short string, and let the item definition select this. An alternative could be for the >item definition to define the splines themselves, but this is for future generations.

You are only adding a few pre-set animations for the person to choose from or are you going to add the ability to use pre-set or a custom one?

**Comment by LoneWolfHT** [comment]
@bendeutsch 
Sorry. Was busy recently.
By 'person' I meant modder. I'll do my best to use that instead from now on to avoid confusion.
The ability to have custom animations would be nice if you're working on a tool that doesn't use any of the preset ones. I don't think they would mind if it was comp;icated. Mobs are complicated

**Comment by ClobberXD** [comment]
Maybe provide standard animations and also provide an API to register new ones?

**Comment by bendeutsch** [comment]
Oh yes, there definitely should be a way to declare custom animations for tools, but probably *after* this pull request, not in it. Otherwise it'll never be finished.

What we can/should do is make sure this PR leads naturally to one where custom animations can be defined.

**Comment by paramat** [comment]
@bendeutsch have you abandoned this PR?

**Comment by bendeutsch** [comment]
Hi, yeah, I've been thinking about this. My first impulse was to say "no, it's not abandoned", since I never made a conscious decision, but I've been working less and less on Minetest stuff over time. I think it's time to make this hiatus official.

I'm closing this PR; the branch and code will still stay up, if anyone wants to continue, but for now, I won't be doing anything with it.

**Comment by paramat** [comment]
Ok, we'll miss you, you are an excellent contributor.
I'll mark all your PRs as 'adoption needed' unless you object.

**Comment by bendeutsch** [comment]
Yes, "adoption needed" should be fine for the time being. Thanks for everything!

**Pull request closed without merging**